test/exceed_test.gd

```diff
@@ -0,0 +1,279 @@
+			return
+		while game_logic.game_state == Enums.GameState.GameState_PlayerDecision:
+				## TODO: Figure out if it's really necessary to limit ourselves to this decision type
+				# and game_logic.decision_info.type == Enums.DecisionType.DecisionType_ChooseSimultaneousEffect:
+			empty_loop_count = 0  # reset the count each time we actually get into this loop; it is not empty
+			var decision = game_logic.decision_info
+			var player = initiator
+			var player_choices = init_choices
+			if decision.player == defender.my_id:
+				player = defender
+				player_choices = def_choices
+			var choice = player_choices.pop_front()
+			if choice == null:
+				fail_test("Insufficient decisions defined for player %s during strike" % player.my_id)
+				return
+			match game_logic.decision_info.type:
+				Enums.DecisionType.DecisionType_ChooseSimultaneousEffect:
+					assert_true(game_logic.do_choice(player, choice),
+							"%s failed to perform a choice with value %s" % [player, choice])
+				Enums.DecisionType.DecisionType_ChooseToDiscard:
+					assert_true(game_logic.do_choose_to_discard(player, choice),
+							"%s failed to discard cards %s" % [player, choice])
+				Enums.DecisionType.DecisionType_ForceForArmor:
+					assert_true(game_logic.do_force_for_armor(player, choice),
+							"%s failed to discard cards %s for armor" % [player, choice])
+				var decision_type:  # Unknown decision type, just roll with it?
+					if typeof(choice) == Variant.Type.TYPE_ARRAY:
+						fail_test("Attempting to apply array choice %s to a decision of type %s" % [
+								choice, Enums.DecisionType.keys()[decision_type]])
+						return
+					assert_true(game_logic.do_choice(player, choice),
+							"Decision of type %s unhandled by test harness (attempted by player %s, content %s)" % [
+									decision_type, player, choice])
+		## TODO: Does the loop need to wait a tick or two here for the game engine to
+		##     present another decision?
+		# wait_seconds(0.01)
+	if game_logic.game_state == Enums.GameState.GameState_PlayerDecision:
+		fail_test("Insufficient decisions defined for player %s during strike" % game_logic.decision_info.player)
+		return
+
+func execute_strike(initiator, defender, init_card: String, def_card: String,
+		init_ex = false, def_ex = false, init_choices = [], def_choices = []):
+	var init_card_id = -1
+	var init_card_ex_id = -1
+	var def_card_id = -1
+	var def_card_ex_id = -1
+	## TODO: Because all the card IDs are generated internally and at runtime, it is hard
+	##   to pass those IDs in through *_choices
+	if init_card:
+		init_card_id = give_player_specific_card(initiator, init_card)
+		if init_ex:
+			init_card_ex_id = give_player_specific_card(initiator, init_card)
+		do_and_validate_strike(initiator, init_card_id, init_card_ex_id)
+	else:
+		init_card_id = do_and_validate_strike(initiator, -1)  # wild swing
+	## TODO: Why no all_events modification in the initiator block? Is it
```

Comment:
Not really sure what this is asking

Reply:
The original behavior of `execute_strike` includes a return of `all_events`, which is the aggregate contents of `local_game.event_queue` over the duration of `execute_strike`. This return value included all events that happened during the defender's response and the decision-making that ensues, but omits the portion of the event queue that happens during the initiator's attack set. The initiator's attack set calls through `do_and_validate_strike`, which inspects `event_queue` (and therefore *empties* it) without returning its contents or adding a copy to `all_events`, so those events are missing. This might be surprising to anyone who looked at the array of events returned by the original `execute_strike` and assumes that it is a *complete* array of events from the strike.

It was unclear to me whether it was important that any manipulation of `all_events` happens during `execute_strike`. I chose to modify things to leave `get_latest_events()` alone entirely for all of `execute_strike`. Any test that actually cares about its contents can call `get_latest_events` directly and get a complete array of relevant events.

Reply:
I'm going to go with - i probably didn't think about it too much and execute_strike grew organically and I only did what I needed for the tests I wrote.